Data/Primitive/PrimArray.hs

```diff
@@ -1163,6 +1155,21 @@ unST (GHCST.ST f) = f
 runPrimArray m = runST $ m >>= unsafeFreezePrimArray
 #endif
 
+-- | Create an uninitialized array of the given length, apply the function to
+-- it, and freeze the result.
+--
+-- /Note:/ this function does not check if the input is non-negative.
+--
+-- @since FIXME
+createPrimArray
+  :: Prim a => Int -> (forall s. MutablePrimArray s a -> ST s ()) -> PrimArray a
+{-# INLINE createPrimArray #-}
+createPrimArray 0 _ = PrimArray (emptyPrimArray# (# #))
```

Comment:
Note: I have replicated the attempt `createArray` and `createSmallArray` make to share empty `Array#`s and `SmallArray#`s. Let me know if we don't want this on `ByteArray#` for some reason.